### ISSUE TITLE
Custom default materials aren't inherited

### ISSUE THREAD

**Issue by hypirion** [IfcOpenShell/IfcOpenShell#473]
When looking up materials for an object and not finding any, IfcOpenShell currently checks whether the object type has any default material. Do you think it makes sense to check this through inheritance as well? For example, in the current code, IfcWall and IfcWallStandardCase both have the same value, even though IfcWallStandardCase is a subtype of IfcWall.

I guess it's not that important right now as one could enumerate all the different types internally, but it would be handy if #472 is eventually accepted.

**Comment by aothms** [comment]
Yeah I agree. For IFC2X3 this wasn't so much of an issue I guess, as I think this wall standard case was one of the very few occasions where a supertype would be concrete and frequently used. In IFC4 there are a lot more of these 'cases' introduced (for reasons that are beyond me).

I don't think this is feasible to implement in the master / 0.5 branch. As C++ classes are not first order citizens there is not really a means to enumerate subtypes.

In the 0.6 branch, there is a full representation of the schema hierarchy available at runtime. 

https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcparse/IfcSchema.h

`declaration_by_name("IfcWall")->as_entity()` would give you a fairly direct equivalent of what is defined in EXPRESS. It seems though that data for the `subtypes()` [1] function is not actually populated, so you'll have to rebuild that from `supertype()` or I can quickly patch the code generator [2] if you want to work on this.

[1] https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcparse/IfcSchema.h#L309
[2] https://github.com/IfcOpenShell/IfcOpenShell/blob/v0.6.0/src/ifcexpressparser/schema_class.py

**Comment by aothms** [comment]
Hi, are you still considering to look into this?

> I think it should be possible to walk up instead of down here (that was my first thought at least)

Yes, you can probably walk up when the data is requested, or walk down when the data is populated initially and fill the map for subtypes as well. Perhaps you're right and the first might handle specialized materials for subtypes better.

Regardless of that, subtypes() are now populated correctly as well.
